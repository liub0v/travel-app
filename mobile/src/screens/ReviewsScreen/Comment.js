import React, {useMemo} from 'react';
import {
  CommentContainer,
  CommentTextWrapper,
  DateTitle,
  UserAvatar,
  UserContainer,
  UserFirstNameTitle,
  UserInfoContainer,
  UserInfoFirstLineWrapper,
  UserInfoWrapper,
  UserRatingTitle,
} from './CommentInput.style';
import {DEFAULT_PROFILE_IMAGE} from '../../constants/api';
import {dateParser} from '../../services/dataParser';
import {Stars as Rating} from '../../components/Stars/Stars';
import {DynamicText} from '../AdventureScreen/AdventureScreen';

export const Comment = ({item}) => {
  const date = new Date(item?.date);

  const nameTitle = useMemo(() => {
    const firstName = item?.clientID?.profileInfo?.firstName ?? '';
    const lastName = item?.clientID?.profileInfo?.lastName ?? '';
    const name = `${firstName} ${lastName}`;
    return name.trim() || item?.clientID?.userID?.username || 'Anonymous';
  }, [item]);
  return (
    <CommentContainer>
      <UserContainer style={{flexDirection: 'row', flex: 1}}>
        {item?.clientID?.profileInfo?.imageURL ? (
          <UserAvatar source={{uri: item?.clientID?.profileInfo?.imageURL}} />
        ) : (
          <UserAvatar source={DEFAULT_PROFILE_IMAGE} />
        )}
        <UserInfoContainer style={{justifyContent: 'space-between', flex: 1}}>
          <UserInfoFirstLineWrapper
            style={{flexDirection: 'row', justifyContent: 'space-between'}}>
            <UserFirstNameTitle>{nameTitle}</UserFirstNameTitle>
            <DateTitle>{dateParser(date)}</DateTitle>
          </UserInfoFirstLineWrapper>
          <UserInfoWrapper>
            <UserRatingTitle>{item?.rating?.generalRating}</UserRatingTitle>
            <Rating starsNumber={item?.rating?.starsNumber} />
          </UserInfoWrapper>
        </UserInfoContainer>
      </UserContainer>
      <CommentTextWrapper>
        <DynamicText text={item?.comment} lineNumber={3} />
      </CommentTextWrapper>
    </CommentContainer>
  );
};
