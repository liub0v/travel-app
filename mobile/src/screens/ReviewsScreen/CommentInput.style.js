import styled from 'styled-components/native/dist/styled-components.native.esm';
import colors from '../../constants/colors';
import fonts from '../../constants/fonts';
export const CommentInputWrapper = styled.View`
  padding-left: 24px;
  padding-right: 24px;
`;
export const CommentInputContainer = styled.ScrollView`
  flex: 1;
  height: 400px;
`;
export const CriterionRatingContainer = styled.View`
  flex: 1;
  margin-top: 24px;
`;
export const StarsRatingContainer = styled.View`
  flex: 1;
  margin-top: 24px;
`;
export const CriterionRatingWrapper = styled.View`
  flex-direction: row;
  width: 90%;
  justify-content: space-between;
`;
export const CommentTextInput = styled.TextInput`
  background-color: ${colors.white};
  color: ${colors.blue};
  font-family: ${fonts.normal};
  font-size: 14px;
  border-radius: 16px;
  margin-top: 12px;
  padding-left: 24px;
  padding-right: 24px;
  padding-top: 12px;
  padding-bottom: 12px;
`;

export const CommentContainer = styled.View``;
export const UserContainer = styled.View`
  padding-left: 24px;
  padding-right: 24px;
  flex-direction: row;
  flex: 1;
`;
export const UserAvatar = styled.Image`
  width: 40px;
  height: 40px;
  border-radius: 20px;
`;
export const ButtonWrapper = styled.View`
  align-self: flex-end;
  margin-top: 24px;
  width: 30%;
`;
export const UserInfoContainer = styled.View``;
export const UserInfoWrapper = styled.View`
  flex-direction: row;
  align-items: center;
  padding-left: 24px;
  padding-top: 24px;
`;
export const UserInfoFirstLineWrapper = styled.View``;
export const CommentTextWrapper = styled.View`
  margin-top: -12px;
  padding-left: 48px;
`;
export const UserFirstNameTitle = styled.Text`
  font-family: ${fonts.normal};
  font-size: 14px;
  color: ${colors.white};
  padding-left: 12px;
`;
export const RatingValueTitle = styled.Text`
  font-family: ${fonts.bold};
  font-size: 14px;
  color: ${colors.white};
  padding-left: 12px;
`;
export const CriterionTitle = styled.Text`
  font-family: ${fonts.normal};
  font-size: 14px;
  color: ${colors.grey};
  padding-left: 12px;
`;
export const StarsRatingTitle = styled.Text`
  font-family: ${fonts.bold};
  font-size: 14px;
  color: ${colors.white};
  padding-left: 12px;
`;
export const StarsTitle = styled.Text`
  font-family: ${fonts.normal};
  font-size: 14px;
  color: ${colors.grey};
  padding-left: 12px;
`;

export const DateTitle = styled.Text`
  font-family: ${fonts.normal};
  font-size: 14px;
  color: ${colors.grey};
  padding-left: 12px;
`;
export const UserRatingTitle = styled.Text`
  font-family: ${fonts.bold};
  font-size: 14px;
  color: ${colors.white};
  padding-left: 12px;
  padding-right: 12px;
`;
export const StarItem = styled.Image`
  width: 32px;
  height: 32px;
`;
export const StarsWrapper = styled.View`
  flex-direction: row;
  align-items: center;
  justify-content: space-around;
  padding-right: 24px;
  padding-left: 24px;
  margin-bottom: 24px;
`;
