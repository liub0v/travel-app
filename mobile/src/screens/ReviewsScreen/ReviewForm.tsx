import React, {useState} from 'react';
import {commentValidationSchema} from '../../services/validation';
import {
  ButtonWrapper,
  CommentInputWrapper,
  CommentTextInput,
  CriterionRatingContainer,
  StarsRatingContainer,
} from './CommentInput.style';
import {CriterionRating} from '../../components/CriterionRating/CriterionRating';
import {StarsRating} from '../../components/StarsRating/StarsRating';
import colors from '../../constants/colors';
import {Text} from 'react-native';
import {ButtonItem} from '../../components/Buttons/ButtonItem';
import {Formik} from 'formik';

export interface FormValues {
  commentText: string;
  interestingRatingValue: number;
  guideRatingValue: number;
  serviceRatingValue: number;
  priceRatingValue: number;
  starsRatingValue: number;
}

interface ProfileFormProps {
  showCriterionRating?: boolean;
  onSubmitHandler: (values: FormValues) => void;
}
export const ReviewForm: React.SFC<ProfileFormProps> = ({
  showCriterionRating,
  onSubmitHandler,
}) => {
  const [starsRatingValue, setStarsRatingValue] = useState(0);

  return (
    <Formik
      validationSchema={commentValidationSchema}
      initialValues={{
        commentText: '',
        interestingRatingValue: 0,
        guideRatingValue: 0,
        serviceRatingValue: 0,
        priceRatingValue: 0,
        starsRatingValue: 0,
      }}
      onSubmit={async values => {
        await onSubmitHandler({...values, starsRatingValue});
      }}>
      {({
        handleChange,
        handleBlur,
        handleSubmit,
        values,
        errors,
        touched,
        setFieldValue,
      }) => (
        <CommentInputWrapper>
          {showCriterionRating && (
            <CriterionRatingContainer>
              <CriterionRating
                testID={'interesting-rating-value'}
                title={'Interesting'}
                ratingValue={values.interestingRatingValue}
                onValueChangeHandler={(value: number) => {
                  setFieldValue('interestingRatingValue', value);
                }}
              />
              <CriterionRating
                testID="guide-rating-value"
                title={'Guide'}
                ratingValue={values.guideRatingValue}
                onValueChangeHandler={(value: number) =>
                  setFieldValue('guideRatingValue', value)
                }
              />
              <CriterionRating
                testID="service-rating-value"
                title={'Service'}
                ratingValue={values.serviceRatingValue}
                onValueChangeHandler={(value: number) =>
                  setFieldValue('serviceRatingValue', value)
                }
              />
              <CriterionRating
                testID="price-rating-value"
                title={'Price'}
                ratingValue={values.priceRatingValue}
                onValueChangeHandler={(value: number) =>
                  setFieldValue('priceRatingValue', value)
                }
              />
            </CriterionRatingContainer>
          )}

          <StarsRatingContainer>
            <StarsRating
              setStarRating={setStarsRatingValue}
              initStarsNumber={starsRatingValue}
            />
          </StarsRatingContainer>

          <CommentTextInput
            testID="comment-text-input"
            multiline={true}
            placeholderTextColor={colors.grey}
            placeholder={'Type a comment...'}
            onChangeText={handleChange('commentText')}
            onBlur={handleBlur('commentText')}
            value={values.commentText}
            name={'comment'}
          />
          {errors.commentText && touched.commentText && (
            <Text
              testID="validation-error"
              style={{color: colors.red, padding: 6}}>
              {errors.commentText}
            </Text>
          )}
          <ButtonWrapper>
            <ButtonItem
              title={'Send'}
              titleSize={12}
              size={{height: 40, width: 100}}
              handler={handleSubmit}
            />
          </ButtonWrapper>
        </CommentInputWrapper>
      )}
    </Formik>
  );
};
