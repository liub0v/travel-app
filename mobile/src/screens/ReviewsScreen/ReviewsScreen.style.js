import styled from 'styled-components/native/dist/styled-components.native.esm';
import colors from '../../constants/colors';

export const TitleWrapper = styled.View`
  margin-bottom: 48px;
  background-color: ${colors.green};
  color: ${colors.green};
  align-items: center;
  padding: 6px;
`;
