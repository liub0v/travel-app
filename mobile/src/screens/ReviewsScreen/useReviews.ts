import {useMemo} from 'react';
import {currentHotelReviewsSelector} from '../../../redux/selectors/HotelSelectors';
import {currentAdventureReviewsSelector} from '../../../redux/selectors/AdventureSelectors';
import {useSelector} from 'react-redux';

export type useReviewsType = {
  comments: Array<any>;
  isCriterionRating: boolean;
};

export const useReviews: useReviewsType = (type: string) => {
  const isCriterionRating = useMemo(() => type === 'adventure', [type]);
  const commentSelector = () => {
    if (type === 'hotel') {
      return currentHotelReviewsSelector;
    }
    if (type === 'adventure') {
      return currentAdventureReviewsSelector;
    }
  };

  const comments = useSelector(commentSelector())?.reverse();

  return {comments, isCriterionRating};
};
