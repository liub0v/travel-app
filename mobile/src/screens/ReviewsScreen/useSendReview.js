import {useEffect, useState} from 'react';
import {showMessage} from 'react-native-flash-message';
import {getAdventureReview} from '../../../redux/actions/AdventureActions';
import {closeSocket} from '../../../redux/actions/CommentActions';
import {useDispatch} from 'react-redux';

export const useSendReview = onSubmit => {
  const dispatch = useDispatch();
  const [isCommentForm, setIsCommentForm] = useState(true);

  const onSubmitHandler = async ({
    commentText,
    interestingRatingValue,
    guideRatingValue,
    serviceRatingValue,
    priceRatingValue,
    starsRatingValue,
  }) => {
    try {
      await onSubmit(
        starsRatingValue,
        commentText,
        interestingRatingValue,
        guideRatingValue,
        serviceRatingValue,
        priceRatingValue,
      );

      setIsCommentForm(false);
    } catch (error) {
      setIsCommentForm(true);
      showMessage({
        message: error.response?.data,
        type: 'error',
      });
    }
  };
  useEffect(() => {
    dispatch(getAdventureReview());
  }, []);

  useEffect(() => {
    return () => dispatch(closeSocket());
  }, []);

  return {isCommentForm, onSubmitHandler};
};
