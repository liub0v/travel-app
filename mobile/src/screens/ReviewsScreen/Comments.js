import React from 'react';
import {Comment} from './Comment';
import {CommentsContainer} from './ReviewsScreen.style';

export const Comments = React.memo(({comments}) => {
  return (
    <CommentsContainer style={{flex: 1}}>
      {comments?.map(item => (
        <Comment item={item} key={item._id} />
      ))}
    </CommentsContainer>
  );
});
