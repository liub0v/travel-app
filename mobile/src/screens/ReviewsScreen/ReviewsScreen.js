import React, {useCallback, useEffect, useState, useMemo} from 'react';

import {useRoute} from '@react-navigation/native';
import {
  UserAvatar,
  UserFirstNameTitle,
  UserInfoWrapper,
} from './CommentInput.style';
import {useDispatch, useSelector} from 'react-redux';
import {
  roleSelector,
  userSelector,
} from '../../../redux/selectors/UserSelector';
import {showMessage} from 'react-native-flash-message';
import {getAdventureReview} from '../../../redux/actions/AdventureActions';

import {currentHotelReviewsSelector} from '../../../redux/selectors/HotelSelectors';
import {currentAdventureReviewsSelector} from '../../../redux/selectors/AdventureSelectors';
import {closeSocket} from '../../../redux/actions/CommentActions';
import {DEFAULT_PROFILE_IMAGE} from '../../constants/api';

import {Container, TitleWrapper} from './ReviewsScreen.style';
import {NormalText} from '../AuthScreens/LoginScreen/LoginScreen.style';
import {Comments} from './Comments';

import {ReviewForm} from './ReviewForm';

export const Author = ({nameTitle, imageURL}) => {
  return (
    <UserInfoWrapper>
      {imageURL ? (
        <UserAvatar source={{uri: imageURL}} />
      ) : (
        <UserAvatar source={DEFAULT_PROFILE_IMAGE} />
      )}
      <UserFirstNameTitle>{nameTitle}</UserFirstNameTitle>
    </UserInfoWrapper>
  );
};
export const ReviewsScreen = () => {
  const route = useRoute();
  const dispatch = useDispatch();

  const type = route.params?.type;
  const onSubmit = route.params?.onSubmit;

  const isCriterionRating = useMemo(() => type === 'adventure', [type]);
  const [isCommentForm, setIsCommentForm] = useState(true);

  const commentSelector = useCallback(() => {
    if (type === 'hotel') {
      return currentHotelReviewsSelector;
    }
    if (type === 'adventure') {
      return currentAdventureReviewsSelector;
    }
  }, [type]);

  const user = useSelector(userSelector);
  const role = useSelector(roleSelector);
  const comments = useSelector(commentSelector())?.reverse();

  const nameTitle = useMemo(
    () =>
      user?.profileInfo?.firstName
        ? user?.profileInfo?.firstName
        : user?.profileInfo?.lastName
        ? user?.profileInfo?.lastName
        : user?.userID?.username,
    [user?.profileInfo?.firstName, user?.profileInfo?.lastName],
  );

  const onSubmitHandler = async ({
    commentText,
    interestingRatingValue,
    guideRatingValue,
    serviceRatingValue,
    priceRatingValue,
    starsRatingValue,
  }) => {
    try {
      await onSubmit(
        starsRatingValue,
        commentText,
        interestingRatingValue,
        guideRatingValue,
        serviceRatingValue,
        priceRatingValue,
      );
      setIsCommentForm(false);
    } catch (error) {
      setIsCommentForm(true);
      showMessage({
        message: error.response?.data,
        type: 'error',
      });
    }
  };

  useEffect(() => {
    dispatch(getAdventureReview());
  }, []);

  useEffect(() => {
    return () => dispatch(closeSocket());
  }, []);

  return (
    <Container>
      {role !== 'admin' && isCommentForm && (
        <>
          <Author
            nameTitle={nameTitle}
            imageURL={user?.profileInfo?.imageURL}
          />
          <ReviewForm
            showCriterionRating={isCriterionRating}
            onSubmitHandler={onSubmitHandler}
          />
        </>
      )}
      {!isCommentForm && (
        <TitleWrapper>
          <NormalText>Thank for your review!</NormalText>
        </TitleWrapper>
      )}
      <Comments comments={comments} />
    </Container>
  );
};
