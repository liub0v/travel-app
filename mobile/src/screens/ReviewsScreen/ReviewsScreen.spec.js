import React from 'react';
import {shallow} from 'enzyme';
import {useRoute} from '@react-navigation/native';
import {useDispatch, useSelector} from 'react-redux';
import {Author, ReviewsScreen} from './ReviewsScreen';
import {STORE} from '../../tests/__mocks__/store-mock';
import {USER_MOCKS} from '../../tests/__mocks__/user-mocks';
import {ReviewForm} from './ReviewForm';
import {Comments} from './Comments';
import {useReviews} from './useReviews';
import {useSendReview} from './useSendReview';

jest.mock('./useReviews');
jest.mock('./useSendReview');

jest.mock('@react-navigation/native');

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

describe('Testing ReviewScreen component', () => {
  let mockEffect;
  let wrapper;
  let mockMemo;
  let mockCallback;

  let user = {
    profileInfo: USER_MOCKS.PROFILE_INFO,
    userID: USER_MOCKS.USER_INFO,
  };
  let reviews = STORE.hotel.currentHotel.data.reviews;
  beforeEach(() => {
    useRoute.mockReturnValue({
      params: {
        type: 'hotel',
        onSubmit: () => jest.fn(() => Promise.resolve({})),
      },
    });
    useReviews.mockReturnValue({
      comments: reviews,
      isCriterionRating: false,
    });
    useSendReview.mockReturnValue({
      isCommentForm: true,
      onSubmitHandler: jest.fn(),
    });

    mockEffect = jest.spyOn(React, 'useEffect');
    mockEffect.mockImplementation(f => f());

    mockCallback = jest.spyOn(React, 'useCallback');
    mockCallback.mockImplementation(f => f());

    mockMemo = jest.spyOn(React, 'useMemo');
    mockMemo.mockImplementation(f => f());

    useDispatch.mockImplementation(() => () => {});
    useSelector
      .mockReturnValueOnce(user)
      .mockReturnValueOnce(user.userID.role)
      .mockReturnValueOnce(reviews);

    wrapper = shallow(<ReviewsScreen />);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should not show review form it role is admin ', () => {
    // useSelector
    //   .mockReturnValueOnce(USER_MOCKS.ADMIN)
    //   .mockReturnValueOnce(USER_MOCKS.ADMIN.userID.role)
    //   .mockReturnValueOnce(reviews);
    useSendReview.mockReturnValue({
      isCommentForm: false,
      onSubmitHandler: jest.fn(),
    });
    wrapper = shallow(<ReviewsScreen />);

    expect(wrapper.find(ReviewForm)).not.toExist();
    expect(wrapper.find(Author)).not.toExist();
    expect(wrapper.find(Comments)).toExist();
  });
  it('should not show review when comment has been send ', async () => {
    await wrapper.find(ReviewForm).invoke('onSubmitHandler')({
      commentText: 'text',
      interestingRatingValue: 0,
      guideRatingValue: 0,
      serviceRatingValue: 0,
      priceRatingValue: 0,
      starsRatingValue: 0,
    });

    expect(wrapper.find(ReviewForm).prop('onSubmitHandler')).toHaveBeenCalled();

    // expect(wrapper.find(ReviewForm)).not.toExist();
    // expect(wrapper.find(TitleWrapper)).toExist();
  });
});
