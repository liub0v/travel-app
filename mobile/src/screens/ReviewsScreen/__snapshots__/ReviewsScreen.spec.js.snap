// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ReviewScreen component should match snapshot 1`] = `
<Styled(ScrollView)>
  <Author
    imageURL="https://cdn.pixabay.com/photo/2016/08/08/09/17/avatar-1577909_960_720.png"
    nameTitle="FirstName"
  />
  <ReviewForm
    onSubmitHandler={[MockFunction]}
    showCriterionRating={false}
  />
  <Memo()
    comments={
      Array [
        Object {
          "_id": 12,
          "clientID": Object {
            "profileInfo": Object {
              "address": "City, Country",
              "birthDate": "November 23, 2001",
              "firstName": "FirstName",
              "imageURL": "https://cdn.pixabay.com/photo/2016/08/08/09/17/avatar-1577909_960_720.png",
              "lastName": "LastName",
              "phone": "80291234567",
            },
          },
          "comment": "comment1",
          "date": 2022-02-11T15:02:47.793Z,
          "rating": Object {
            "generalRating": 9,
            "starsNumber": 4,
          },
          "userID": Object {
            "email": "user@mail.com",
            "role": "client",
            "username": "username",
          },
        },
        Object {
          "_id": 13,
          "clientID": Object {
            "profileInfo": Object {
              "address": "City, Country",
              "birthDate": "November 23, 2001",
              "firstName": "FirstName",
              "imageURL": "https://cdn.pixabay.com/photo/2016/08/08/09/17/avatar-1577909_960_720.png",
              "lastName": "LastName",
              "phone": "80291234567",
            },
          },
          "comment": "comment2",
          "date": 2022-02-11T15:02:47.793Z,
          "rating": Object {
            "generalRating": 8,
            "starsNumber": 3,
          },
          "userID": Object {
            "email": "user@mail.com",
            "role": "client",
            "username": "username",
          },
        },
        Object {
          "_id": 14,
          "clientID": Object {
            "profileInfo": Object {
              "address": "City, Country",
              "birthDate": "November 23, 2001",
              "firstName": "FirstName",
              "imageURL": "https://cdn.pixabay.com/photo/2016/08/08/09/17/avatar-1577909_960_720.png",
              "lastName": "LastName",
              "phone": "80291234567",
            },
          },
          "comment": "comment3",
          "date": 2022-02-11T15:02:47.793Z,
          "rating": Object {
            "generalRating": 10,
            "starsNumber": 5,
          },
          "userID": Object {
            "email": "user@mail.com",
            "role": "client",
            "username": "username",
          },
        },
      ]
    }
  />
</Styled(ScrollView)>
`;
